core/shared/src/main/scala/zio/prelude/fx/ZPure.scala

```diff
@@ -541,21 +553,25 @@ sealed trait ZPure[+W, -S1, +S2, -R, +E, +A] { self =>
   final def repeatWhileState(f: S2 => Boolean)(implicit ev: S2 <:< S1): ZPure[W, S1, S2, R, E, A] =
     repeatUntilState(!f(_))
 
+  /**
+   * Runs this computation to produce its result.
+   */
+  final def run(implicit ev1: Unit <:< S1, ev2: Any <:< R, ev3: E <:< Nothing): A =
+    runResult(())
+
   /**
    * Runs this computation with the specified initial state, returning both
    * the updated state and the result.
    */
   final def run(s: S1)(implicit ev1: Any <:< R, ev2: E <:< Nothing): (S2, A) =
-    runEither(s).fold(ev2, identity)
-
-  final def runEither(s: S1)(implicit ev: Any <:< R, ev1: CanFail[E]): Either[E, (S2, A)] =
-    runEitherCause(s).fold(cause => Left(cause.first), Right(_))
+    runAll(s)._2.fold(cause => ev2(cause.first), identity)
 
   /**
-   * Runs this computation with the specified initial state, returning either a
-   * failure or the updated state and the result
+   * Runs this computation with the specified initial state, returning both the
+   * log and either all the failures that occurred or the updated state and the
+   * result.
    */
-  final def runEitherCause(s: S1)(implicit ev: Any <:< R, ev1: CanFail[E]): Either[Cause[E], (S2, A)] = {
+  final def runAll(s: S1)(implicit ev: Any <:< R): (Chunk[W], Either[Cause[E], (S2, A)]) = {
```

Comment:
I find it counter-intuitive that `Chunk[W]` is returned _even if_ the computation finished successfully.

I thought that `ZPure` is now also a replacement for `Validated` (or was it named `Validation`?) which aggregates multiple errors (instead of bailing on the first one).

What should _ZIO Prelude_'s story be for something like this?

```scala
def parseInt(s: String): Validation[ParseError, Int] = ???

val list = List("1", "2", "hello", "3", "world")
val parsed = list.traverse(parseInt)

List(ParseError(...), ParseError(...))
```

---

I'm also wondering, whether we need the distinction between `W` and `E`. Could we have just one error type `E` and then have two methods on ZPure:
 * one would accumulate all the errors and gave you `Either[Chunk[E], A]
 * the other would bail on first encountered error and returned `Either[E, A]`

:thinking:

Reply:
If multiple errors occur they are captured in the `Cause` which reflects the full structure of parallel and sequential errors and can be exposed using operators like `Sandbox`. I don't think we can combine `W` and `E` because `W` is a log that we want include in the success case as well, for example we might just be logging what we are doing so we can go back and look at the steps we took later.

Reply:
You can get the two different behaviors you are talking about by using `foreach` (fails fast) versus `foreachPar` (accumulates errors) and then using one of the operators that exposes all the errors that occurred.

Reply:
Oh, I've misunderstood the purpose of `W` then. Thanks for the explanation :+1:

Reply:
Yeah, I realize there was a lot that went into that one PR. Originally there was `Validation` and then we wanted to augment it to be `ZValidation` with the log parameter and that got me thinking about unifying with `ZPure`. But yeah `W` is a writer monad effect which could be used for logging with different interpretations of what you have done or potentially a list of warnings that were triggered but didn't prevent the computation from continuing, and then `E` through the `Cause` structure handles the accumulation of errors and more ability to retain the structure of what order they occurred in than a traditional non-empty collection.